File: runners/csharp/Kikker.Billing.IntegrationTests/Helpers/ComposeCommandRunner.cs
namespace Kikker.Billing.IntegrationTests.Helpers;

public sealed class ComposeCommandRunner
{
    private readonly BillingTestOptions options;

    public ComposeCommandRunner(BillingTestOptions options)
    {
        this.options = options;
    }

    public async Task<string> RunAsync(params string[] args)
    {
        var startInfo = new ProcessStartInfo("docker")
        {
            WorkingDirectory = options.RepositoryRoot,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false
        };

        startInfo.ArgumentList.Add("compose");
        startInfo.ArgumentList.Add("--env-file");
        startInfo.ArgumentList.Add(options.EnvFile);
        startInfo.ArgumentList.Add("-f");
        startInfo.ArgumentList.Add(options.ComposeFile);

        foreach (var arg in args)
        {
            startInfo.ArgumentList.Add(arg);
        }

        using var process = new Process { StartInfo = startInfo };
        process.Start();

        var stdout = process.StandardOutput.ReadToEndAsync();
        var stderr = process.StandardError.ReadToEndAsync();

        await process.WaitForExitAsync();

        var output = await stdout;
        var error = await stderr;

        if (process.ExitCode != 0)
        {
            throw new InvalidOperationException($"docker compose command failed with exit code {process.ExitCode}.{Environment.NewLine}STDOUT:{Environment.NewLine}{output}{Environment.NewLine}STDERR:{Environment.NewLine}{error}");
        }

        return output;
    }
}

File: runners/java/src/test/java/com/kikker/billing/integrationtests/helpers/ComposeCommandRunner.java
package com.kikker.billing.integrationtests.helpers;

import com.kikker.billing.integrationtests.BillingTestOptions;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public final class ComposeCommandRunner {
    private final BillingTestOptions options;

    public ComposeCommandRunner(BillingTestOptions options) {
        this.options = options;
    }

    public String run(String... args) throws IOException, InterruptedException {
        List<String> command = new ArrayList<>();
        command.add("docker");
        command.add("compose");
        command.add("--env-file");
        command.add(options.envFile().toString());
        command.add("-f");
        command.add(options.composeFile().toString());

        for (String arg : args) {
            command.add(arg);
        }

        Process process = new ProcessBuilder(command)
            .directory(options.repositoryRoot().toFile())
            .start();

        String stdout = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
        String stderr = new String(process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
        int exitCode = process.waitFor();

        if (exitCode != 0) {
            throw new IllegalStateException(
                "docker compose command failed with exit code " + exitCode + System.lineSeparator()
                    + "STDOUT:" + System.lineSeparator() + stdout + System.lineSeparator()
                    + "STDERR:" + System.lineSeparator() + stderr
            );
        }

        return stdout;
    }
}
